[export] add runtime asserts on replacement expressions

pianpwk opened the pull request:
Currently, Hybrid SymInts can lead to symbols being replaced with nonlinear expressions of other symbols. For example, the following reshape + add emits a guard `s0*s1 == s2`, which replaces `s2` with `s0*s1`:
```
def forward(self, x, y):
    # x: [s0, s1]; y: [s2]
    return x.flatten() + y
```
The exported program then looks like this, with `s0*s1` containing its own ValueRange:
```
ExportedProgram:
    class GraphModule(torch.nn.Module):
        def forward(self, w: "f32[s0, s1]", x: "f32[s0*s1]"):
            view: "f32[s0*s1]" = torch.ops.aten.view.default(w, [-1]);  w = None
            add: "f32[s0*s1]" = torch.ops.aten.add.Tensor(view, x);  view = x = None
            return (add,)

Range constraints: {s0: VR[0, int_oo], s1: VR[0, int_oo], s0*s1: VR[0, int_oo]}
```

During analysis, the runtime assert `Eq(s0*s1, s2)` is added, which results in this `_set_replacement()` call. However, during the runtime asserts pass, `s2` no longer exists in the program as an input symbol, and so the runtime assert never gets reified. This means that nothing exists in the program to assert this equality, and an error is raised during the op instead.

This PR proposes to look up ShapeEnv replacements, and makes sure to reify the replaced symbol with a `sym_size` or `sym_stride` call, and computes the replacement expression from other input shapes, to assert equality.

ezyang commented:
This is failing tests, do you want it reviewed now or later? While I find it plausible there's some sort of hybrid symint problem, I find the description on the PR a bit impenetrable.

ezyang commented on torch/fx/passes/runtime_assert.py:
```diff
@@ -251,11 +257,21 @@ def match_symbol(symint, cb):
+                            # if we've set a replacement, e.g. s2 = s0*s1, s0*s1 shows up as the shape,
+                            # so we store expr_to_proxy[s2] = cb(), and we're able to reify the assert.
+                            expr_to_proxy[_s] = fx.Proxy(cb())
+                            log.debug("expr_to_proxy[%s] = %s", _s, expr_to_proxy[_s])
```
This doesn't feel tight. If you want, I'll think more about what a tight way of doing this, but this is probably what I'd think about on how to express it in a more general way.

pianpwk commented on torch/export/_trace.py:
```diff
@@ -103,7 +103,7 @@ class ExportDynamoConfig:
     # but if we want to reason more about what guards/runtime asserts to emit,
     # this makes it a bit cleaner to do from the export side. Also no real point in running this twice.
-    do_not_emit_runtime_asserts = True
+    do_not_emit_runtime_asserts: bool = True
```
apparently this didn't get picked up, so _dynamo.export() was still running the runtime asserts pass for strict

pianpwk commented on test/export/test_export.py:
```diff
@@ -6401,20 +6390,6 @@ def forward(self, x, y):
-            if is_non_strict_test(self._testMethodName)
-            and not is_training_ir_test(self._testMethodName)
-            else 0,
-        )
```
after modifying the flag below, dynamo doesn't DCE this op, and there's now no strict/non-strict mismatch

pianpwk committed:
[export] add runtime asserts on replacement expressions (#131051)

Summary:
Currently, Hybrid SymInts can lead to symbols being replaced with nonlinear expressions of other symbols. For example, the following reshape + add emits a guard `s0*s1 == s2`, which replaces `s2` with `s0*s1`:
```
def forward(self, x, y):
    # x: [s0, s1]; y: [s2]
    return x.flatten() + y
```
The exported program then looks like this, with `s0*s1` containing its own ValueRange:
```
ExportedProgram:
    class GraphModule(torch.nn.Module):
        def forward(self, w: "f32[s0, s1]", x: "f32[s0*s1]"):
            view: "f32[s0*s1]" = torch.ops.aten.view.default(w, [-1]);  w = None
            add: "f32[s0*s1]" = torch.ops.aten.add.Tensor(view, x);  view = x = None
            return (add,)

Range constraints: {s0: VR[0, int_oo], s1: VR[0, int_oo], s0*s1: VR[0, int_oo]}
```

During analysis, the runtime assert `Eq(s0*s1, s2)` is added, which results in this `_set_replacement()` call. However, during the runtime asserts pass, `s2` no longer exists in the program as an input symbol, and so the runtime assert never gets reified. This means that nothing exists in the program to assert this equality, and an error is raised during the op instead.

This PR proposes to look up ShapeEnv replacements, and makes sure to reify the replaced symbol with a `sym_size` or `sym_stride` call, and computes the replacement expression from other input shapes, to assert equality.

Pull Request resolved: https://github.com/pytorch/pytorch/pull/131051

Differential Revision: D59867179

Pulled By: pianpwk

pianpwk commented:
> This is failing tests, do you want it reviewed now or later? While I find it plausible there's some sort of hybrid symint problem, I find the description on the PR a bit impenetrable.

@ezyang I just updated the PR + description, would you mind re-reviewing?

Alternatively, maybe there's a way to not replace symbols with expressions in placeholder shapes? At least on the export side, there's no current benefit to having a symbolic shape like `[s0*s1, ...]` in a placeholder - we don't do anything special for it, and it hurts in cases like these.

ezyang commented:
We have to deal with an analogous problem when we have replacements for backed symbols. The algorithm here is that for every input to our program, we generate a guard that specifies that this input must equal the symbolic expression we allocated for it. If tensor.size(0) == s2 (and s2 is not mentioned anywhere else) this is trivial, but if tensor.size(0) == s0 * s1 now nontrivial action happens.

I was expecting something analogous to happen here; but now it has to happen in the insert runtime asserts pass, rather than guard generation. You need to go over all hybrid symint arguments and generate runtime asserts for their replacements. I would NOT rely on residual runtime assert in the list of asserts (i.e., s0 * s1 == s2), because you can easily get in trouble if some "simplifies" this expression to (s0 * s1 == s0 * s1) (after all, you have the replacement!) and now the runtime assert is a no-op. I think we're a little less aggressive about this sort of thing on deferred runtime asserts, but it is better to avoid this problem entirely.

cc @lezcano

ezyang commented:
I think another valid fix would also be to NEVER to do replacements on hybrid symints. I can support a fix like that too.

pianpwk commented:
Closing this, moving to https://github.com/pytorch/pytorch/pull/133920 instead
